### ISSUE TITLE
Authoritative server can report overflowed ttl (near UINT_MAX)

### ISSUE THREAD

**Issue by smikes** [PowerDNS/pdns#5603]
- Program: Authoritative, Recursor
 - Issue type: Bug report

### Short description

When the authoritative server is forwarding requests to the recursive server, it caches responses for TTL seconds.  However, it appears there is a bug that causes TTL for some records to be decremented in two situations

- once per second, as expected
- once per request, incorrectly

I see responses like this (partial dig output)
```
cname.test.		4294967292 IN	CNAME	target.test.
```

The attached tar.gz file contains instructions for reproducing this behavior using dockerized CentOS 6.

### Environment

 - Operating system: CentOS 6
 - Software version: 4.0.6
 - Software source: PowerDNS repository

### Steps to reproduce
Extract tarball to centos-pdns-repro on a machine where you can run docker.
```
cd centos-pdns-repro
docker build -t centos-pdns-repro .
docker run centos-pdns-repro /tmp/repro.sh
```

### Expected behaviour
I expect a dns server to never return a TTL which is greater than the TTL of the source record.  In this configuration, the CNAME has TTL 120, but is reported with TTL approaching INT_MAX.

### Actual behaviour
Output of the repro script:
```
Starting tinydns:                                          [  OK  ]
Starting pdns-recursor:                                    [  OK  ]
Starting PowerDNS authoritative nameserver: started
Priming caches and waiting 1 second
Starting to query pdns (authoritative)
TTL=119 TTL=118 TTL=117 TTL=116 TTL=115 TTL=114 TTL=113 TTL=112 TTL=111 TTL=110 TTL=109 TTL=108 TTL=107 TTL=106 TTL=105 TTL=104 TTL=103 TTL=102 TTL=101 TTL=100 TTL=99 TTL=98 TTL=97 TTL=96 TTL=95 TTL=94 TTL=93 TTL=92 TTL=91 TTL=90 TTL=89 TTL=88 TTL=87 TTL=86 TTL=85 TTL=84 TTL=83 TTL=82 TTL=81 TTL=80 TTL=79 TTL=78 TTL=77 TTL=76 TTL=75 TTL=74 TTL=73 TTL=72 TTL=71 TTL=70 TTL=69 TTL=68 TTL=67 TTL=66 TTL=65 TTL=64 TTL=63 TTL=62 TTL=61 TTL=60 TTL=59 TTL=58 TTL=57 TTL=56 TTL=55 TTL=54 TTL=53 TTL=52 TTL=51 TTL=50 TTL=49 TTL=48 TTL=47 TTL=46 TTL=45 TTL=44 TTL=43 TTL=42 TTL=41 TTL=40 TTL=39 TTL=38 TTL=37 TTL=36 TTL=35 TTL=34 TTL=33 TTL=32 TTL=31 TTL=30 TTL=29 TTL=28 TTL=27 TTL=26 TTL=25 TTL=24 TTL=23 TTL=22 TTL=21 TTL=20 TTL=19 TTL=18 TTL=17 TTL=16 TTL=15 TTL=14 TTL=13 TTL=12 TTL=11 TTL=10 TTL=8 TTL=6 TTL=4 TTL=2 TTL=0 TTL=4294967294
Reproduced the bug after 116 queries

; <<>> DiG 9.8.2rc1-RedHat-9.8.2-0.62.rc1.el6 <<>> @127.0.0.1 -p 2053 a cname.test
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 26585
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;cname.test.			IN	A

;; ANSWER SECTION:
cname.test.		4294967292 IN	CNAME	target.test.
target.test.		86276	IN	A	127.0.0.2

;; Query time: 0 msec
;; SERVER: 127.0.0.1#2053(127.0.0.1)
;; WHEN: Tue May 23 21:04:08 2017
;; MSG SIZE  rcvd: 65
```

[centos-pdns-repro.tar.gz](https://github.com/PowerDNS/pdns/files/1218383/centos-pdns-repro.tar.gz)

**Comment by pieterlexis** [comment]
Thank you for the detailed report and reproduction scenario. The recursive component of the authoritative server is known to have 'interesting' corner cases when it comes to caching. The recursive component is [removed](https://doc.powerdns.com/authoritative/changelog/4.1.html#change-da5e9c3c24b7b1060858aa9e6d58a3bb) from the authoritative server in the next release (4.1.0).

**Comment by ahupowerdns** [comment]
So what happens is that recursive cached packets in 4.0, it did not correctly extract the minimum TTL from packets received from the recursing backend. Since we've deprecated this feature in 4.1, we are not going to fix it in 4.0. For 4.0, I recommend setting `recursive-cache-ttl=0` which will effectively solve the issue for you, without a big slowdown (unless you are operating at really high query rates). Sorry we can't do better than this, but we do thank you for your high quality debugging!
